# Basic Ensemble Tecnique
## Import Libraries

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


# Read Dataset

In [4]:
dataset = pd.read_csv("placement.csv")
dataset.head()

cgpa  package
0  6.89     3.26
1  5.12     1.98
2  7.82     3.25
3  7.42     3.67
4  6.94     3.57

# Separte Indep and Dep var

In [6]:
x = dataset.iloc[:, :-1]
y = dataset["package"]

In [7]:
x

cgpa
0    6.89
1    5.12
2    7.82
3    7.42
4    6.94
..    ...
195  6.93
196  5.89
197  7.21
198  7.63
199  6.22

[200 rows x 1 columns]

In [8]:
x.ndim

2

# Train Test Split

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
x_train, x_test, y_train, y_test = train_test_split(x, y , test_size=0.2, random_state=42)

# Three Models Use

In [13]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR

# LinearRegression

In [15]:
# Model
lr = LinearRegression()
lr.fit(x_train, y_train)

# Accuracy scores
train_score = lr.score(x_train, y_train) * 100
test_score = lr.score(x_test, y_test) * 100

# Print with labels
print(f"Training Accuracy: {train_score:.2f}%")
print(f"Testing Accuracy: {test_score:.2f}%")

Training Accuracy: 77.58%
Testing Accuracy: 77.31%


# DecisionTreeRegressor

In [17]:
# Model
dt = DecisionTreeRegressor()
dt.fit(x_train, y_train)

# Accuracy scores
train_score = dt.score(x_train, y_train) * 100
test_score = dt.score(x_test, y_test) * 100

# Print with labels
print(f"Training Accuracy: {train_score:.2f}%")
print(f"Testing Accuracy: {test_score:.2f}%")

Training Accuracy: 95.16%
Testing Accuracy: 52.20%


# SVR

In [19]:
# Model
sv = SVR()
sv.fit(x_train, y_train)

# Accuracy scores
train_score = sv.score(x_train, y_train) * 100
test_score = sv.score(x_test, y_test) * 100

# Print with labels
print(f"Training Accuracy: {train_score:.2f}%")
print(f"Testing Accuracy: {test_score:.2f}%")

Training Accuracy: 78.17%
Testing Accuracy: 75.95%


# Voting Regressor

In [21]:
from sklearn.ensemble import VotingRegressor

In [22]:
li = [("lr1", LinearRegression()), ("dt1", DecisionTreeRegressor()), ("sv1", SVR())]

In [23]:
vc = VotingRegressor(li)    # weight also use weights = [10, 20, 30]  for weightage averaging
vc.fit(x_train, y_train)

VotingRegressor(estimators=[('lr1', LinearRegression()),
                            ('dt1', DecisionTreeRegressor()), ('sv1', SVR())])

In [24]:
# Accuracy scores
train_score = vc.score(x_train, y_train) * 100
test_score = vc.score(x_test, y_test) * 100

# Print with labels
print(f"Training Accuracy: {train_score:.2f}%")
print(f"Testing Accuracy: {test_score:.2f}%")

Training Accuracy: 87.57%
Testing Accuracy: 75.17%


# Show Prediction that Acutually work or not

In [26]:
prd = { "Linear Regression" : lr.predict(x_test) ,
        "Decision Tree":  dt.predict(x_test) ,
        "SVM" : sv.predict(x_test),
        "Voting Regressor": vc.predict(x_test)}

In [27]:
pd.DataFrame(prd).head(5)


Linear Regression  Decision Tree       SVM  Voting Regressor
0           2.780313         2.9200  2.745462          2.815259
1           3.136352         3.2525  3.111375          3.166742
2           3.199521         2.9200  3.186923          3.102148
3           2.389819         2.5400  2.407183          2.445667
4           3.526847         3.9900  3.603909          3.706919

In [28]:
# yes it Acutually Work
(2.780313+2.9200+2.745462)/3

2.815258333333333

# Ensemble Learning with Voting Regressor

We combined multiple regression models into a **Voting Regressor** to improve prediction stability and accuracy.

---

## 🔧 Models Used

- **Linear Regression**  
- **Decision Tree Regressor**  
- **Support Vector Regressor (SVR)**  
- **Voting Regressor** (Averaging method)

---

## 🔍 Prediction Comparison

| Linear Regression | Decision Tree | SVM Regressor | Voting Regressor |
|-------------------|---------------|---------------|------------------|
| 2.780313          | 2.9200        | 2.745462      | 2.815259         |
| 3.136352          | 3.2525        | 3.111375      | 3.166742         |
| 3.199521          | 2.9200        | 3.186923      | 3.102148         |
| 2.389819          | 2.5400        | 2.407183      | 2.445667         |
| 3.526847          | 3.9900        | 3.603909      | 3.706919         |

👉 The **Voting Regressor** output is calculated by averaging predictions from all base models.

---

